winch/codegen/src/visitor.rs

```diff
@@ -1189,13 +1189,39 @@ where
             self.context.pop_to_reg(self.masm, None)
         };
 
-        self.masm.branch(
-            IntCmpKind::Ne,
-            top.reg.into(),
-            top.reg.into(),
-            *frame.label(),
-            OperandSize::S32,
-        );
+        // Emit instructions to balance the machine stack if the frame has
+        // a different offset.
+        let current_sp_offset = self.masm.sp_offset();
+        let (_, frame_sp_offset) = frame.base_stack_len_and_sp();
+        if current_sp_offset == frame_sp_offset {
+            self.masm.branch(
+                IntCmpKind::Ne,
+                top.reg.into(),
+                top.reg.into(),
+                *frame.label(),
+                OperandSize::S32,
+            );
+        } else {
+            let br_fallthrough_label = self.masm.get_label();
+            self.masm.branch(
+                IntCmpKind::Eq,
+                top.reg.into(),
+                top.reg.into(),
+                br_fallthrough_label,
+                OperandSize::S32,
+            );
```

Comment:
It seems that we can avoid some repetition here, as only the label and the `CmpKind` are different for the branch instruction:

```rust
let (label, cmp, needs_cleanup) = if current_sp_offset > frame_sp_offset {
    (masm.get_label(), IntCmpKind::Eq, true)
} else {
    (•frame.label(), IntCmpKind::Ne, false)
};

self.masm.branch(...);

if needs_cleanup { .. } 
```